src/transformStateWithClones.js

```diff
@@ -7,7 +7,36 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const tempState = { ...state };
+  const statesArr = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(tempState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const rem of action.keysToRemove) {
+          delete tempState[rem];
+        }
+        break;
+
+      case 'clear': {
+        for (const key in tempState) {
+          delete tempState[key];
+        }
+        break;
+      }
+
+      default:
+        break;
+    }
+
+    statesArr.push({ ...tempState });
```

Comment:
It's better to push a copy of the object to the array at the end of each loop cycle, but outside of the `switch` block, after the current action is processed. This is already done correctly here.